### DISCUSSION TITLE
How to get live feedback of formatting errors as I type?

### DISCUSSION THREAD

**Discussion by zorzysty** [biomejs/biome#1684] [Q&A]
I'd like to see, right inside my IDE, whenever there's a formatting issue that will be resolved by running Biome's formatter - similarly to how I see linting errors as I type. Is that possible using Biome's plugin for VS Code (and perhaps Intellij)?

Background: in my current project we have a prettier plugin for ESLint which informs us about formatting issues via ESLint. This is obviously very slow, but as a team, we love this functionality, and we'd like to keep it after migrating from Prettier to Biome.

Awesome project BTW!

**Comment by castarco** [comment]
Hi @zorzysty 

I'm not sure that Biome identifies formatting problems as "problems", as far as I know, it just fixes them without "complaining".

What you can do, though, to ensure that formatting is as much automated as possible, is to configure VSCode to run the formatter at specific times (for example, when you save, or a few seconds after you stop tipping).

For example, if you open VSCode settings, and type one of the following search queries (you can easily adapt them to affect just the current project):
```
@id:editor.defaultFormatter @lang:json format
```
```
@id:editor.defaultFormatter @lang:javascript format
```
```
@id:editor.defaultFormatter @lang:typescript format
```

You will see something like:

<img width="1187" alt="Screenshot 2024-04-12 at 17 59 55" src="https://github.com/biomejs/biome/assets/251364/1c9fc91c-7d5e-432d-b27e-d61bce8616bb">

There you can set Biome as your default formatter for the specific language you typed in your search query.

Note that you can also configure _when_ the formatter will execute:

<img width="1187" alt="Screenshot 2024-04-12 at 18 02 01" src="https://github.com/biomejs/biome/assets/251364/6ea468c4-a099-45cb-9621-be5ee80344b3">

**Comment by ltianyi992** [comment]
Yes — the **Biome VS Code extension** supports live formatting error display, but you need to enable it explicitly:

**Setup in VS Code:**

1. Install the [Biome extension](https://marketplace.visualstudio.com/items?itemName=biomejs.biome)

2. In your VS Code settings (`.vscode/settings.json`):

```json
{
  "editor.defaultFormatter": "biomejs.biome",
  "editor.formatOnSave": true,
  
  // Show formatter diffs as diagnostics (yellow squiggles) as you type:
  "biome.lspBin": "node_modules/.bin/biome",
  
  // Enable format-on-save so issues resolve on save
  "[javascript]": {
    "editor.defaultFormatter": "biomejs.biome",
    "editor.formatOnSave": true
  },
  "[typescript]": {
    "editor.defaultFormatter": "biomejs.biome",
    "editor.formatOnSave": true
  }
}
```

3. Make sure Biome is installed as a project dependency:
```bash
npm install --save-dev @biomejs/biome
```

**How live formatting feedback works:**
The Biome LSP server (started by the extension) analyzes your file on change and emits diagnostics for formatting issues. These show as squiggles in the editor with a "This code should be reformatted" message — the same feedback you would get from `biome format --write` but inline.

This is similar to the prettier-plugin-eslint approach but much faster (Biome's LSP is Rust-based and responds in < 1ms in most cases).

**JetBrains/IntelliJ:** The official [Biome IntelliJ plugin](https://plugins.jetbrains.com/plugin/22761-biome) also shows formatting diagnostics inline via the LSP integration — install it from the plugin marketplace and point it to your project's Biome binary.
